### ISSUE TITLE
Downgrade WARNING: PolySet -> Manifold conversion failed

### ISSUE THREAD

**Issue by kintel** [openscad/openscad#5135]
We currently issue a warning if someone tries to specify a polyhedron which does not represent a manifold topology:
```
WARNING: PolySet -> Manifold conversion failed: NotManifold
Trying to repair and reconstruct mesh..
```

This warning feels a bit intimidating, and it also disrupts a bunch of existing designs and libraries (see e.g. discussion in #5134).

We should consider downgrading or silencing this warning in cases where merging co-incident vertices yields a valid manifold without having to repair/reorient the mesh.

Example: Closed extrusion where the first and last cross section is repeated:
```
polyhedron(points = [
  [2.02196e-17, 19.8, 1],
  [-2.02196e-17, 20.2, 1],
  [-2.02196e-17, 20.2, -1],
  [2.02196e-17, 19.8, -1],
  [10.8034, -1.52112, 1],
  [11.16, -1.70232, 1],
  [11.16, -1.70232, -1],
  [10.8034, -1.52112, -1],
  [-10.8034, -1.52112, 1],
  [-11.16, -1.70232, 1],
  [-11.16, -1.70232, -1],
  [-10.8034, -1.52112, -1],
  [2.02196e-17, 19.8, 1],
  [-2.02196e-17, 20.2, 1],
  [-2.02196e-17, 20.2, -1],
  [2.02196e-17, 19.8, -1]
],
  faces = [[0, 5, 4],
  [0, 1, 5],
  [1, 2, 5],
  [5, 2, 6],
  [2, 3, 6],
  [6, 3, 7],
  [3, 0, 7],
  [7, 0, 4],
  [4, 5, 8],
  [8, 5, 9],
  [5, 6, 9],
  [9, 6, 10],
  [6, 7, 10],
  [10, 7, 11],
  [7, 4, 11],
  [11, 4, 8],
  [8, 9, 12],
  [12, 9, 13],
  [9, 10, 13],
  [13, 10, 14],
  [10, 15, 14],
  [10, 11, 15],
  [11, 8, 15],
  [15, 8, 12]
],
  convexity = 2);
```

**Comment by adrianVmariano** [comment]
Is there a disadvantage to relying on OpenSCAD to merge duplicate vertices?  (The one obvious disadvantage would be that it could in some circumstance produce an incorrect or invalid result.)   I think I favor the idea of an informational comprehensible message that says "Duplicate vertices in model were merged" or something to that effect vs just being completely silent about the matter.

**Comment by kintel** [comment]
As of today, it's not just a question of duplicating vertices. If a non-manifold topology was specified we'll do a bunch of stuff to try to ingest it:
* Quantize vertices to a grid
* Triangulate all polygons
* If the geometry was convex, try to apply a convex hull
* reorient faces which were pointing the wrong way (could also happen due to quantization)

We could split that up into a naive vertex merge first, and if that fails, issue said warning and try to repair.

**Comment by adrianVmariano** [comment]
Are non-triangular polygons considered a fault?  This has always been a point of uncertainty for us, as it seems like there are rare cases where non-triangular coplanar polygons fail, but usually they are OK.   (I suppose there is the question about how accurately coplanar a non-triangular polygon is.)

**Comment by kintel** [comment]
Non-triangular polygons are OK, but there's always a chance that a polygon isn't planar. Triangulating ensures that. ..plus Manifold requires triangles.

**Comment by adrianVmariano** [comment]
What happens if you feed it a non-coplanar polygon?

**Comment by kintel** [comment]
We'll just make a decision for you how to combine the vertices into triangles.

**Comment by DougPollard** [comment]
I was worried when @kintel first mentioned the steps that would be taken if the polyhedron was non-manifold and wondered if it would be more efficient to just eliminate the duplicate vertices in BOSL. So I ran a simple test on a project I'm working on right now. I am generating a closed polyhedron via path_sweep() and vnf_polyhedron(), similar to my original example code, except much closer to planar in appearance. Then I merge a number of these into a nxn mesh to create a weave pattern. For now the result is basically planar but eventually this mesh itself will be wrapped around 3d shapes. Anyways, for a few values of n I ran this, accepting the current Manifold warnings, but successful rendering; and I ran it again except inserting vnf_merge_points() around each result of path_sweep() before passing to vnf_polyhedron. Every test was run after flushing caches.
```  
       n       BOSL        Manifold
       6        40s            4.5s
       7        54s            5.5s
       8        70s            7.5s
       9        90s            9.5s
```
n of 6 means 36 polyhedrons merged, 7 is 49, etc. My finely calibrated eye says there's a linear relation between polyhedron count and time and, more  importantly, Manifold appears consistently about 10x faster at eliminating the duplicates (and rendering) than getting BOSL to do the elimination first. BTW, when I eliminate the duplicates with vnf_merge_points the warning messages are gone.

**Comment by adrianVmariano** [comment]
The OpenSCAD language is horrible for efficient implementation of complicated algorithms due to the impossibility to (efficiently) implement data structures like priority queue that are essential to those algorithms.   It's always going to be **much** faster for an operation to happen in OpenSCAD natively than to run in userspace.   

The point merge process in BOSL2 attempts to speed the search by organizing the data into a tree.  If this works really well then the point merge process will be O(n log n) but if it works less well it can still be O(n^2).  But the speed of this operation definitely depends on how the data is distributed.  

In your test results above, you list a BOSL2 time and a Manifold time, but I'm not sure I know what the times are exactly.   To know how long manifold takes to deal with repeated vertices you'd need a manifold time with and without those repeated vertices.  Did you do a test like that somehow?  Because manifold is doing other stuff, not *only* dealing with manifold repair.

**Comment by DougPollard** [comment]
Yeah, I agree with all points. My test was really just a basic determination, from the user perspective, of which approach is currently best -- ask BOSL2 to clean up the duplicate vertices or let Manifold. The numbers were so overwhelming that I stopped there. It would be nice though, for the developers as well as users, to optimize the task when possible.

I'm not sure when cacheing comes in to play and how to defeat it in OpenSCAD. If we create a polyhedron with some duplicate vertices, then replicate that polyhedron 100 times with some slight transform, does OpenSCAD optimize (remove duplicate vertices) one copy of the polyhedron then reuse that result for the other 99? Or is the cache only used for identical data from one run to the next?

If cacheing is not a concern, then maybe the simple tests are

1. one polyhedron with duplicates removed by BOSL2 and 99 copies merged (ie, Manifold doesn't  have to remove duplicates and the price paid for BOSL2 to do it is small, relative to the overall size of the render)
2. 100 copies merged of polyhedron with duplicates (ie, Manifold has to do the removal 100 times)
3. various ratios of clean-to-dirty, like 50 clean polyhedrons (from one single BOSL2 action) and 50 copies of non-clean, etc

This might provide some insight into how costly the duplicate removal process is in Manifold. It wouldn't answer the question of if that price varies by the number of duplicates per polyhedron or if it's relative constant, based simply on the overall size of the polyhedron. You've already indicated in BOSL2 that that same price does vary based on the specific data, but I'm not sure if that is based on the quantity of duplicates or on how well the order of the vertices fill a tree.

Where I'm heading with this long ramble is maybe there is a price to pay each time Manifold encounters a polyhedron with any duplicates, regardless of the quantity, and so it might behoove BOSL2 to avoid duplicates in certain "easy" cases but not to bother in hard cases. Lots of speculation at this point (that's probably wrong anyways!). Maybe I'll setup a test like this, unless someone tells me it's obviously wrong or meaningless.

**Comment by kintel** [comment]
Minor clarification: The mentioned repair isn't something done by Manifold, it's something OpenSCAD does to be able to pass valid meshes to Manifold.

Caching isn't properly implemented yet for Manifold. That's something we need to resolve before moving Manifold out of the experimental stage: https://github.com/openscad/openscad/issues/4825

**Comment by adrianVmariano** [comment]
With regards to BOSL2 cleaning up duplicate vertices, there are places where it could be done with no significant computational cost---"just" programming effort.  The case of closed path_sweep is probably one such example, though there are some subtleties (specifically, if the object twists, the alignment between the vertices is not necessarily a direct mapping, which I think would actually be a major complication).  The performance issue with vnf_merge_points arises from relying on a brute force approach that doesn't exploit topology.  But it's a big change and a lot of effort to change things to exploit toplogy---the obvious thing would be to track where the boundaries of partial polyhedra are so that when they are merged, the points on the boundary are lined up and merged.  This would be faster since it reduces the points that need to be checked to a presumably small boundary set.  (In principle you could align one point and then follow the boundary, but that might be hard to implement efficiently.)   There is definitely a question of whether any of these things are worth putting effort toward.  I don't think we'd want to try to do the boundary tracking until we get real structures.  

For testing purposes, you could test path sweeps with closed=true where you invoke vnf_merge_points or don't, and look at **preview** time difference, which should tell you how long BOSL2 takes to remove duplicate vertices.  Then without invoking vnf_merge_points you can compare the **render** time for closed=true and closed=false.  Since closed=false has no duplicate points but closed=true does have duplicate points, this should be a pretty good test of the render time required to remove the duplicate points to prepare the input for manifold.

**Comment by nophead** [comment]
If have just come across this warning in my threads library. I have two separate polyhedrons that are the starts and ends of the thread and I intersect them with a rotate_extruded polygon to chamfer or truncate them. I don't think there are any duplicated points in my polyhedra. Either operand of the intersection is fine on its own but together they give this warning.

**Comment by adrianVmariano** [comment]
Does the rule still apply that a single polyhedron doesn't invoke the geometry engine?   That is, if you take one polyhedron and combine it with a cube do you get the warning?

**Comment by nophead** [comment]
The ends get unioned with the rest of the thread and that doesn't generate any warnings if I don't intersect them first.

**Comment by nophead** [comment]
I do get the warning if I union the ends with a cube.

**Comment by nophead** [comment]
The intersection is inside a render(). If I remove that I don't get the warning. I also only get 5 warnings for 8 threads of different sizes.

**Comment by nophead** [comment]
This is all in a preview. I never do a render of these, so I suppose the intersection is done by OpenCSG. It is only when I add the render it is done by manifold.

**Comment by kintel** [comment]
@nophead the warning is most likely a real indication that your mesh isn't a manifold topology and requires some sort of repair or vertex merge. I suggest you open a separate ticket and post the smallest example you can come up with, and we can help identify the root cause.

**Comment by nophead** [comment]
Wouldn't CGAL barf if I did an intersection with non-manifold polyhedra? Without manifold selected the rendered intersection works fine. The ends of the thread are just a simple sweep.

![image](https://github.com/openscad/openscad/assets/566149/5e6e91fe-373f-4cb7-9069-774ff8985025)

This is what the intersection looks like.

![image](https://github.com/openscad/openscad/assets/566149/01afe119-14ff-4c33-aa93-22653829e81c)

**Comment by nophead** [comment]
It was a simple case of somebody passing a thread profile with a zero crest. That caused a duplicate point in the profile. Sorry for the noise.

**Comment by kintel** [comment]
> Does the rule still apply that a single polyhedron doesn't invoke the geometry engine?   That is, if you take one polyhedron and combine it with a cube do you get the warning?  

It's more of an implementation detail. We convert to the geometry backend-specific format only when needed, and it's currently not needed for rendering a single polyhedron. You can force it to do this on the cmd-line for testing: openscad file.scad -o out.png --render=force

**Comment by kintel** [comment]
> It was a simple case of somebody passing a thread profile with a zero crest. That caused a duplicate point in the profile. Sorry for the noise.

If you include the duplicate vertices in the mesh with their own indices, this should now work: The topology is manifold and degenerate triangles caused by duplicate vertex positions are tolerated. If you find odd corner cases feel free to provide an example :)

**Comment by adrianVmariano** [comment]
Do you mean duplicate vertices that have (degenerate) triangles that place them consistently into the mesh?  Or do you mean duplicate vertices where the mesh assumes that the duplicate vertices are the same point even though they have different indices?

**Comment by kintel** [comment]
The first.

**Comment by adrianVmariano** [comment]
Has something changed about how duplicate vertices get treated?   I'm not sure if this is related to the issue discussed here or is a separate matter, but someone reported an artifact that arose because two objects were unioned but they had a 1e-15 gap between them and this gap ended up appearing in the final model and lead to a render failure.   The same model in the stable version rendered successfully.

**Comment by kintel** [comment]
If the issue isn't about gaps in a single polyhedron it's worth opening a new ticket.

**Comment by kintel** [comment]
> The case of closed path_sweep is probably one such example, though there are some subtleties (specifically, if the object twists, the alignment between the vertices is not necessarily a direct mapping, which I think would actually be a major complication).

@adrianVmariano In this specific case, it sounds like you're relying on merging _almost_ coincident vertices. That's an important detail to keep in mind,

**Comment by adrianVmariano** [comment]
Yeah, it's possible in the specific case of twisting path_sweep that round-off error in the rotation would mean the vertices wouldn't be exactly coincident, but off by machine epsilon.  Will the duplicate vertex handling work in this case?

The situation described above involved 2d polygons with a 1e-15 error that was a result of a rotation, and it actually turns out that it happened in both versions of OpenSCAD but for some reason was easier to see in the dev version.

**Comment by kintel** [comment]
It should work, but we may end up merging close vertices more aggressively than expected. We should come up with some nomenclature for how we deal with such geometry so we can document what you can expect.
I might end up splitting up the current repair functionality into multiple steps, and only issue a warning if the "regular" steps fail and we go into attempting a full mesh reorientation.

**Comment by adrianVmariano** [comment]
I think in most cases, duplicated vertices will be exact duplicates.  The twisted path_sweep is one exception.   I can imagine that someone building up some shape with beziers could get points differing by machine epsilon if they compute different patches that theoretically share the same boundary---and that's an example of a use case where it would be very difficult for us to change things to prevent the duplication of points.

**Comment by adrianVmariano** [comment]
So this means that identical duplicate vertices produce no message of any kind, and almost identical vertices work as before, producing a standard warning?

**Comment by kintel** [comment]
Exactly.

**Comment by kintel** [comment]
..but be aware that there is a subset of manifold objects not representable after merging vertices, things like:
* Two cubes touching each other
* Donut with a zero radius hole
* ..or any self-touching objects.

**Comment by adrianVmariano** [comment]
It appears that this still hasn't been acted upon, as someone complained in the forum that their render failed in BOSL2 due to this issue.

**Comment by kintel** [comment]
Note that this still only produces a warning - it should have no impact on the geometry creation
Can you provide a minimal breakage example?

**Comment by adrianVmariano** [comment]
As I previously noted, OpenSCAD is unusable unless "stop on first warning" is enabled, so yes, getting a warning blocks geometry creation.   The person who raised the issue on the forum had enabled "stop on first warning".  Also I don't think that there is currently any warning you can get that shouldn't actually be an error.  This is the first such warning that I am aware of.   Whenever my code stops on first warning the code is wrong or has bugs.   Gibberish like `translate(["yes","hello","four"]) cube();` is only a warning.  

Here is a short example of the failure to create geometry due to round-off error in point locations:

```
include <BOSL2/std.scad> 
include <BOSL2/rounding.scad>

union(){
  rounded_prism(square(10), height=10, joint_top=1);
  cuboid(3);
}
```

Result:

```
Parsing design (AST generation)...
Compiling design (CSG Tree generation)...
Rendering Polygon Mesh using Manifold...
WARNING: PolySet -> Manifold conversion failed: NotManifold
Trying to repair and reconstruct mesh..
Rendering cancelled on first warning.
WARNING: No top level geometry to render
```

**Comment by kintel** [comment]
ok, let's try to further evaluate in which cases we can repair such geometry without losing data.
I'll add this ticket as a release blocker for now.

**Comment by adrianVmariano** [comment]
The kinds of geometry that BOSL2 creates should involve well-behaved duplication.  They arise in this case because mating bezier patches are computed separately and then placed into a single polyhedron, so the shared edge has point pairs that don't exactly agree.  At corners there will be more duplicates, but there shouldn't be anything weird with the topology, or any points that are extremely close but need to be kept separate to maintain the topology.

**Comment by kintel** [comment]
The devil is in the details: My current concern is that going through CGAL will remove all colors. The code paths used by BOSL most likely don't support colors, so this is probably not a concern for now. However, the internal code paths used to pass data to manifold _does_ support colors. This means that we need to be a bit careful about how we perform such repairs, as well as which warnings should be considered informational and which indicate that data may be lost.

**Comment by adrianVmariano** [comment]
How does/will color support work?  

Would there be a way to supply a polyhedron fixer that we could call from userspace to handle specific situations, e.g. to locate duplicate points in a polyhedron?  This is expensive to do in userspace.

We can also consider ways in the future to track edges and merge them in userspace, which I think may be possible and would substantially reduce the performance penalty, but I don't want to attempt that until we have user-creatable structures.   This would completely eliminate the problem for situations where we are just gluing edges together, which is maybe the only problem case?   (I'm not sure if any other problem cases exist, but nothing comes to mind.)

**Comment by kintel** [comment]
Colors are per-triangle colors, but creation or queries of individual triangle colors are currently not exposed to user space, but are managed internally when doing CSG operations, import, export etc. of objects with colors.

Offering user space functions on 3D objects are a bit far off, so it's not in scope for this discussion.

**Comment by adrianVmariano** [comment]
I don't need user space 3d objects.  I just need a function that takes a list of 3d points and identifies the approximate duplicates.  If that operation was fast then we could just do it any time we have made a polyhedron with duplicated edges.   Or there could be an option on polyhedron() to merge points that are closer than a specified epsilon.

**Comment by nophead** [comment]
I get this warning for this bit of simple code. A difference of a cube from a rotate_extrude of slightly more than 180 degrees. I suspect the end of the extrusion is not quite planar due to floating point inaccuracies but it is certainly a valid manifold.

```
hob_r = 3 / 2;
hob_R = 5 / 2;
hob_crest = 0.1;
hob_teeth = 50;
hob_pulley_d = 12;
hob_depth = 1;

eps = 1 /128;
$fa = 1;
$fs = 0.1;

hob_pulley_r = hob_pulley_d / 2;
tooth_w = 2 * hob_pulley_r * sin(180 / hob_teeth);

render() difference() {
    rotate_extrude(angle = 180 + eps)
        hull() {
            translate([hob_r, -hob_crest / 2])
                square([eps, hob_crest]);

            translate([hob_R - eps, -tooth_w / 2])
                square([eps, tooth_w]);
        }

    cube([2 * hob_R + 1, 2 * (hob_r - hob_depth), tooth_w + 1], center = true);
}
```

The code works fine with CGAL and if I remove the + eps and do exactly 180 it removes the manifold warning. I can't remember why it was there in the first place as the cube removes the ends. 

This is the shape it produces.

![image](https://github.com/user-attachments/assets/8fb7c4b4-a04f-4ade-90c2-fc30d325ac02)

It is used to add the teeth on this hobbed pulley.

![image](https://github.com/user-attachments/assets/303ee602-7631-4635-9ade-ceac0c636f42)

**Comment by jordanbrown0** [comment]
Here's a simpler demonstration of Chris's case.  I adjusted eps to the close to the largest value that triggers the problem.

```
eps = 0.0094;
$fn = 3;

// Use a triangle to ensure that it's always planar.
module triangle() {
    polygon([
        [0,0],
        [1,0],
        [0,1]
    ]);
}

rotate_extrude(angle = 180 + eps)
    triangle();

translate([10,0,0]) cube(1);
```

The cube is needed to trigger CGAL/Manifold; it isn't otherwise important.

It looks like this is indeed tied to floating point accuracy.  Looking at the STL generated if I drop the cube and so avoid the problematic conversion, the face at the "end" of the conversion is
```
      vertex -0.9999999865420024 -0.00016406094895147938 0
      vertex 0 0 0
      vertex 0 0 1
```
... which is just a tiny bit counterclockwise of the -X axis.

On the other hand, the two faces that are supposed to mate with it at that corner are
```
      vertex 0 0 0
      vertex -0.9999999865420024 -0.00016406094895197543 0
      vertex -0.0000820304747519792 0.9999999966355007 0
```
and
```
      vertex -0.0000820304747519792 0.9999999966355007 0
      vertex -0.9999999865420024 -0.00016406094895197543 0
      vertex 0 0 1
```

Note that the Y coordinate of the corner of the end face is  -0.00016406094895147938, while the Y coordinate of that corner of the other two faces is -0.00016406094895197543, very slightly different.  This isn't shocking, because it's calculated differently,
https://github.com/openscad/openscad/blob/756e080c7e49072d9926cf9ce766def180a0dcae/src/geometry/GeometryEvaluator.cc#L868
versus
https://github.com/openscad/openscad/blob/756e080c7e49072d9926cf9ce766def180a0dcae/src/geometry/GeometryEvaluator.cc#L887

I would have expected the grid snap to fix this difference up - it's at about 10^-15 - but I don't know how that really works.

It seems to me that there may be two appropriate fixes.

First, the rotate_extrude implementation could use exactly the same calculation for the "end" face that it uses for the "ring" faces - or, better, don't recalculate at all but rather use the results from the "ring" calculation for the "end" face.  (In an OpenSCAD polyhedron, I would think of calculating the points once and then reusing the indexes.)  It would be best to do the same for the "start" face too.

Second, my understanding is that this is exactly the sort of thing that the grid snap is supposed to address, so there's the question of why the grid snap isn't working.

**Comment by jordanbrown0** [comment]
Of course, in my analysis above I am assuming that the ASCII STL representation of the polyhedron is an accurate-enough representation of the internal polyhedron.

**Comment by jordanbrown0** [comment]
Note that after
https://github.com/openscad/openscad/blob/756e080c7e49072d9926cf9ce766def180a0dcae/src/geometry/GeometryEvaluator.cc#L884
rings[0] will have the correct coordinates for the start face, and after the loop from 885 through 904 one of the two ring[] vectors - `ring[fragments%2]`, I think - will have the correct coordinates for the end face.

**Comment by jordanbrown0** [comment]
Give or take any need to chop the ends up into triangles.

**Comment by kintel** [comment]
This looks like a bug in rotate_extrude, which was exposed by the manifoldness check, and I've opened a new issue for this: #5555

This is actually a good example why this warning can be useful, and why it's important that we don't remove it until we understand valid vs. invalid corner cases.

**Comment by nophead** [comment]
Why did it work with CGAL? Is the grid snap only done when CGAL is used?

**Comment by kintel** [comment]
CGAL always used grid snap, and with Manifold we try to avoid it unless needed to repair non manifold objects

**Comment by adrianVmariano** [comment]
Just got a complaint about a failed render because of this message.  Is there a path forward here where we don't have a warning message when the point merge repair is performed?

**Comment by adrianVmariano** [comment]
I can't control what my users do and I think most people are using the dev version to get manifold because it is so much faster than CGAL.  This situation can result in users filing bug reports against BOSL2.   I can tell them to turn off "stop on first warning" but as already noted, it's very annoying to run OpenSCAD without "stop on first warning" enabled, since this can result in cascading messages, possibly hundreds of lines of output.  Almost all "warnings" are actually errors and render everything downstream into nonsense.

**Comment by RolandasRazma** [comment]
@adrianVmariano sorry, didn't check your previous messages

**Comment by kintel** [comment]
@nophead I believe your issues related to rotate_extrude usage will be fixed as of #5764

**Comment by kintel** [comment]
Does anyone have a trivial, non-BOSL, test triggering this issue? Ideally one that shows how vertices get generated.

**Comment by adrianVmariano** [comment]
If you just mean the problem of coalescing inexact duplicate vertices producing a warning then this:

v = [[0, 0, 0], [10, 0, 0], [10, 10, 0], [0, 10, 0], [0, 0, 10], [10, 0, 10], [10, 10, 10], [0, 10, 10], [0,0,1e-14]];
f = [[0, 1, 2, 3], [8, 4, 5, 1], [1, 5, 6, 2], [2, 6, 7, 3], [3, 7, 4, 0], [6, 5, 4, 7]];
polyhedron(v, faces=f);
translate([5,5,5])cube(10);

That example is maybe worse behaved than what BOSL2 makes.  In the BOSL2 case the polyhedron is made from toplogically consistent pieces and the problem is just where they are glued together along boundary lines.

**Comment by kintel** [comment]
Thx, that's a good start. The next test I was hoping for was a demo of how a polygon with a transformation ends up giving inaccurate results when the "gluing together" happens. I assume under the hood, the inaccuracy results from applying a series of transformations?

**Comment by kintel** [comment]
The main challenge I'm chipping away at is how to distinguish valid attempts at gluing together two polygons which should ideally be identical vs. detecting unexpected manifoldness issues resulting from internal or external bugs.

**Comment by adrianVmariano** [comment]
Yes, the issue arises when transformations introduce rounding error along the seam where the shape is being "glued together", and yes, I think this is generally the result of a series of transformations.  

Do you want an example that is programmatically produced vs an example that's just a list of vertices and faces?  If I can find a simple example I might be able to de-BOSL-ify it.

**Comment by kintel** [comment]
A programmatically produced example would be great. Perhaps start with a minimal BOSL example for reference, then extract a polyhedron literal from it, and later look into de-boslifying.

**Comment by jordanbrown0** [comment]
Remember that two identical vec3s, transformed through identical chains of transformations, will yield identical results.

The problems are when the input vec3s are mathematically identical but in practice slightly different because they were constructed differently, or when the chains of transformations are mathematically equivalent but in practice slightly different.

**Comment by kintel** [comment]
@jordanbrown0 Yes, this discussion is about the latter, e.g. where a 2D polygon is swept through 3D space until it touches itself. The start- and end vertices end up being identical within some epsilon caused by successive transformations.

**Comment by adrianVmariano** [comment]
```
include<BOSL2-fork/std.scad>
include<BOSL2-fork/isosurface.scad>

v_size = 8;
bbox = [[-16,-17,25], [-1,7,35]];

front_leg = metaballs([ 
    move([-9,-4,30]) * zrot(30) * scale([1.5,5,1.75]), mb_sphere(3),
    move([-9,10,30]), mb_sphere(2, negative = true),
], bbox, v_size, show_stats = true);

vnf_polyhedron(front_leg)
cube(21,anchor=FWD+BACK+BOT);
```

There's an example that is failing, reduced to a minimal polyhedron.  It's produced through a complicated method that in principle produces identical vertices.   The polyhedron has 16 triangular faces and 48 vertices.    The metaball code produces a triangle soup, so I think every triangle has its own set of vertices.

**Comment by adrianVmariano** [comment]
I can't seem to create the problem (any more?) with path_sweep(), but I have realized that triggering the error seems kind of sensitive to where I place my second object.  

```
   rounded_prism(pentagon(3), height=3,
                 joint_top=0.5, joint_bot=0.5, joint_sides=0.5) position(FWD) cube(1);
```

triggers the problem, but I  missed it at first with the cube elsewhere.    The rounded_prism method works by constructing bezier patches for every edge, corner, and face and then combining them, so the edges are produced from separate computations.

**Comment by jordanbrown0** [comment]
Sweeping and meeting yourself seems like it should be straightforward. Just make sure you use the original transform at the end, rather than calculating it anew. (Eg use the 0-degree transform rather than the 360-degree transform.)

But I think Adrian's problems are more with stitching together Bézier patches.

**Comment by adrianVmariano** [comment]
The starting incident for this issue (which was a previous issue) was actually a path_sweep.    Since then I have modified path_sweep to do what you say, which is somewhat more complex because there may be twist involved, so actually the transformation isn't identical.  But it still works anyway.  At least for all the cases I tried, with the cubes where I put them.  

I tried also not letting on that I was making a closed shape so it didn't use that special logic and it still worked.

**Comment by kintel** [comment]
FYI: I'm working on a solution in #5765 which should hopefully allow us to resolve most of the BOSL-related issues before resorting to brute-force repair and issuing a warning.

**Comment by adrianVmariano** [comment]
Not sure if you need more cases, but I found a different class of cases that trigger this problem, which is the textured linear_sweep in BOSL2.   Here's a minimal example:


```
include<BOSL2/std.scad>
rotate_sweep([[10,-10],[10,10]], texture="diamonds",closed=false,tex_reps=[3,1]);  
cuboid(19);
```

This was the smallest reasonable case with the simplest texture, repeated the fewest times, but it seems to fail reliably with any input.

**Comment by kintel** [comment]
@adrianVmariano If you're able to build from source, please give this a try: #5765 
I've added tests representing your latest examples, and it looks pretty robust from my side.

**Comment by adrianVmariano** [comment]
I have not tried building from source, though I assume I could do it.  I probably won't have time to attempt that for several days, though.

**Comment by adrianVmariano** [comment]
I built from source for the first time ever and have been using the build actually to try to chase down a non-manifold bug issue which has been really confusing because the shapes almost always issue the warning.   So far I haven't found a case that works in CGAL and fails in the PR build.  

There is one thing that's a bit annoying, though, which is everything is 10x slower than the "official" nightly build 2025.02.25.snap.  Any idea why this might be the case?  Preview is slower as well as render, so I imagine it's compile settings of some kind.

**Comment by nophead** [comment]
Have you build with a debug option on? When I last used C++ from visual studio it was about 10 times slower in debug mode because it doesn't inline anything.

**Comment by adrianVmariano** [comment]
I built on Linux and followed the instructions provided which didn't say anything about enabling/disabling a debug setting.

**Comment by kintel** [comment]
Perhaps try with -DCMAKE_BUILD_TYPE=Release

**Comment by kintel** [comment]
But basically, what I hear you saying is that the PR behaves better than the current master? In that case I‘ll prepare to merge

**Comment by adrianVmariano** [comment]
I haven't done what I'd call exhaustive testing, but on probably a dozen somewhat complex (but also kind of related) models it eliminated the spurious warning.

**Comment by kintel** [comment]
#5765 is now merged and should make it into the next development snapshot.
I'll leave this ticket open for a bit in case more corner cases are discovered.

**Comment by kintel** [comment]
No further corner cases have been raised. Closing for now.
